### ISSUE TITLE
Unexpected env builtin behaviour

### ISSUE THREAD

**Issue by Cenness** [jqlang/jq#3149]
**Describe the bug**
If  you run jq inside a script with env.VAR and VAR is a bash variable, which wasn't `export`ed before use, jq will use `null` as a value.  
If you happen to not use that variable as a key, jq will exit with no errors.

**To Reproduce**
```
#!/bin/bash

unset k
unset v

k=key
v=value

echo 'using args - works'
jq --arg k "$k" \
   --arg v "$v" \
   -n '.[$k]=$v'

echo 'using env. - fails'
jq -n '.[env.k]=env.v'

echo 'using env. in value - works'
jq -n '.hardkey=env.v'

echo 'using env. with exported vars - works'
export k=key
export v=value
jq -n '.[env.k]=env.v'
```
output:
```
using args - works
{
  "key": "value"
}

using env. - fails
jq: error (at <unknown>): Cannot index null with null

using env. in value - works
{
  "hardkey": null
}

using env. with exported vars - works
{
  "key": "value"
}
```

**Expected behavior**
env should fail if an unset variable is referenced, i.e - tell me that `k` and `v` is undefined in this case

**Environment (please complete the following information):**

- Debian 12
- jq 1.7.1

**Additional context**
N/A

**Comment by Cenness** [comment]
I guess someone somewhere has a valid use case with a `null` as a possible input that should be passed to json query. More flags?

**Comment by wader** [comment]
Hello, I think most of this boils down to how object index and assignment works in jq. `env` is a function that return the current process environment as an object. If you index an object in jq with a string that does not exist it will result in `null`, if the key is not a string you get an error.

```
$ jq -n env
{
  "TERM_SESSION_ID": "<...>",
  <...>
}
```

> ```
> using args - works
> {
>   "key": "value"
> }
> ```

This work because `null` is a bit special in jq, it will act a "empty" value, for assignment this will be `{}`.

```sh
$ jq -cn 'null | .+123, .+"abc", .a = 123'
123
"abc"
{"a":123}
```

> ```
> using env. - fails
> jq: error (at <unknown>): Cannot index null with null

Here you will end doing:
```sh
$ jq -cn 'null | .[null] = null'
jq: error (at <unknown>): Cannot index null with null
```
as `k` and `v` do not exist as keys for `env`, and assigning using `null` as key is an error.

> ```
> using env. in value - works
> {
>   "hardkey": null
> }
> ```

now the key is not `null` so it works, and the value is `null` as it's missing in `env`

> ```
> using env. with exported vars - works
> {
>   "key": "value"
> }
> ```

here the `env` object has both `k` and `v` so it works

> I guess someone somewhere has a valid use case with a null as a possible input that should be passed to json query. More flags?

Not sure i follow here, could you explain more in detail?

**Comment by Cenness** [comment]
Hi. My question isn't about why it works or not. I thought my examples demonstrated that.

It's about failing more loudly. I do not want to accidentally set value to null or sieve through a huge filter to find which var I forgot to export.

To help me to find my mistakes faster I want env to tell me explicitly that referenced var is null, i.e. when I call
```
jq -n '.hardkey=env.v'
```
jq shouldn't happily echo null, but instead fail with 
```
jq: error (at env.v): undefined env var
```

> > I guess someone somewhere has a valid use case with a null as a possible input that should be passed to json query. More flags?
> 
> Not sure i follow here, could you explain more in detail?

This request if implemented would make jq exit with failure on null in env variable. So if someone uses env.SOMETIMES_UNSET_VAR to conditionally write null to a value they would suddenly have broken script after an update. So some kind of flag is needed, either to allow unset vars, or to fail with unset vars.

**Comment by wader** [comment]
> Hi. My question isn't about why it works or not. I thought my examples demonstrated that.
> 
> It's about failing more loudly. I do not want to accidentally set value to null or sieve through a huge filter to find which var I forgot to export.
> 
> To help me to find my mistakes faster I want env to tell me explicitly that referenced var is null, i.e. when I call
> 
> ```
> jq -n '.hardkey=env.v'
> ```
> 
> jq shouldn't happily echo null, but instead fail with
> 
> ```
> jq: error (at env.v): undefined env var

Yes understand. But then either the `env` function would have to return a special object that fails on non-existing keys, or jq has to change how all object indexing works (would be a breaking change). Alternatively add a new CLI option to make all object indexing work that way, `--strict-indexing` etc.

> ```
> 
> > > I guess someone somewhere has a valid use case with a null as a possible input that should be passed to json query. More flags?
> > 
> > 
> > Not sure i follow here, could you explain more in detail?
> 
> This request if implemented would make jq exit with failure on null in env variable. So if someone uses env.SOMETIMES_UNSET_VAR to conditionally write null to a value they would suddenly have broken script after an update. So some kind of flag is needed, either to allow unset vars, or to fail with unset vars.

Yeap if something like `--strict-indexing` would be added it would work like that. The index of non-existing key would throw an error and unless it's caught with `try` it would bubble up and cause a runtime error.

BTW you can implement this in jq already, not so neat but something like this:

```sh
$ k=key v=value jq -n 'def strict_index($k): if has($k) then .[$k] else error("has no key \($k)") end; .[env | strict_index("k")] = (env | strict_index("v"))'
{
  "key": "value"
}

$ k=key jq -n 'def strict_index($k): if has($k) then .[$k] else error("has no key \($k)") end; .[env | strict_index("k")] = (env | strict_index("v"))'
jq: error (at <unknown>): has no key v
```

**Comment by pkoppstein** [comment]
@Cenness wrote:
> This request if implemented would make jq exit with failure on null in env variable.

Please note that when `env.X` returns `null`, that is jq's way of saying that `X` was unset, not that X was set to `"null"`.  Thus the two cases (in bash: `unset X`, and `X=null`) are properly distinguished.
It's also worth noting that in shells such as bash, `echo $X` does not raise an error if X is unset.

Furthermore, as @wader has pointed out, in jq, `env` is just a JSON object.  Thus the idea that env.X should raise an error when X is unset is simply untenable.  

In short, the situation with jq is similar to the situation in bash.  In both cases, if you want to check whether an environment variable has been set, a special mechanism must be used.
